File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        int i = 0;
        unordered_map<char, int> mp;
        int start = 0;
        int ans = 0;
        while(i < n){
            if(mp.find(s[i]) != mp.end()){
                start = mp[s[i]] + 1;
                mp.clear();
                i = start;
            }else{
                mp[s[i]] = i;
                ans = max(ans, i - start + 1);
                i++;
            }
        }
        return ans;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>();

        int l = 0, n = s.length(), ans = 0;
        for(int h = 0; h < n; h++){
            char ch = s.charAt(h);
            while(set.contains(ch) && l < h){
                set.remove(s.charAt(l));
                l++;
            }   
            set.add(s.charAt(h));
            ans = Math.max(h - l + 1,  ans);
        }
        return ans;
    }
}